addons/pos_loyalty/models/loyalty_reward.py

```diff
@@ -11,3 +12,20 @@ def _get_discount_product_values(self):
         for vals in res:
             vals.update({'taxes_id': False})
         return res
+
+    def _should_compute_all_discount_product_pos(self, reward):
+        reward_discount_product_domain = reward._get_discount_product_domain()
+        if not reward_discount_product_domain:
+            return False
+        pos_product_product_loaded_fields = set(
+            self.env['pos.session'].POS_PRODUCT_PRODUCT_FIELDS) | {'id'}
+        for leaf in reward_discount_product_domain:
+            if len(leaf) == 1 and leaf in DOMAIN_OPERATORS:
+                continue
+            if leaf[0] not in pos_product_product_loaded_fields:
+                return True
+        return False
+
+    def _should_compute_all_discount_product(self, reward):
+        return super()._should_compute_all_discount_product(reward) or \
+            self._should_compute_all_discount_product_pos(reward)
```

Comment:
Why did you create a new function?

Reply:
it was in my opinion a cleaner way than doing:
```py
def _should_compute_all_discount_product(self, reward):
  res = super()._should_compute_all_discount_product(reward):
  if res:
     return res
  ... # code of _should_compute_all_discount_product_pos()
```

as or is lazy evaluated, it won't execute the right part if the left one is `True` which is pretty elegant in my opinion.

Let me know if you have another idea that I might have missed

Reply:
> as or is lazy evaluated, it won't execute the right part if the left one is True which is pretty elegant in my opinion.
Doing this in one function achieves the same outcome. 

Since you don't plan to use '_should_compute_all_discount_product_pos' elsewhere, it's better to include your code within the '_should_compute_all_discount_product' function for a neater arrangement.